pkg/gather/gather.go

```diff
@@ -0,0 +1,256 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gather
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/submariner-io/submariner-operator/internal/cli"
+	"github.com/submariner-io/submariner-operator/internal/component"
+	"github.com/submariner-io/submariner-operator/internal/constants"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/brokercr"
+	"github.com/submariner-io/submariner-operator/pkg/client"
+	"github.com/submariner-io/submariner-operator/pkg/cluster"
+	"github.com/submariner-io/submariner-operator/pkg/names"
+	"github.com/submariner-io/submariner-operator/pkg/reporter"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Options struct {
+	Type                 string
+	Module               string
+	Directory            string
+	IncludeSensitiveData bool
+}
+
+const (
+	Logs      = "logs"
+	Resources = "resources"
+)
+
+var ModuleFlags = map[string]bool{
+	component.Connectivity:     false,
+	component.ServiceDiscovery: false,
+	component.Broker:           false,
+	component.Operator:         false,
+}
+
+var TypeFlags = map[string]bool{
+	"logs":      false,
+	"resources": false,
+}
+
+var gatherFuncs = map[string]func(string, Info) bool{
+	component.Connectivity:     gatherConnectivity,
+	component.ServiceDiscovery: gatherDiscovery,
+	component.Broker:           gatherBroker,
+	component.Operator:         gatherOperator,
+}
+
+func Data(clusterInfo *cluster.Info, status reporter.Interface, options Options, writer io.Writer) bool {
+	if options.Directory == "" {
+		options.Directory = "submariner-" + time.Now().UTC().Format("20060102150405") // submariner-YYYYMMDDHHMMSS
+	}
```

Comment:
The codes creates files so it doesn't seem conducive as a library. I think we should keep all this internal.